Add DMA support for SDMMC on STM32F7

msobkowski opened the pull request:
This PR introduces DMA support for SDMMC on STM32F7, along with a fix for
the ext2 filesystem to allow using it with this DMA.

It was tested on the `stm32f746g_disco` board. The SDMMC seems to work fine,
although the DMA in some cases generates an "uexpected irq":

```
[00:00:04.008,000] dma_stm32_v1: FiFo error.
[00:00:04.013,000] dma_stm32_v1: tc: 1, ht: 0, te: 0, dme: 0, fe: 1
[00:00:04.020,000] dma_stm32: Unexpected irq happened.
[00:00:04.027,000] stm32_sdmmc: DMA callback error with channel 3.
```

According to the STM32F7 TRM it does not necessarily mean that data is lost:

> When a FIFO overrun or underrun condition occurs, the data is not lost because
> the peripheral request is not acknowledged by the stream until the overrun or
> underrun condition is cleared. If this acknowledge takes too much time, the
> peripheral itself may detect an overrun or underrun condition of its internal
> buffer and data might be lost.

laudominik committed:
fs: ext2: add alignment for superblock

Reading superblock didn't work with DMAs that only support aligned reads
(e.g. STM32 DMA).

Co-authored-by: Maciej Sobkowski <msobkowski@antmicro.com>
Signed-off-by: Dominik Lau <dlau@internships.antmicro.com>

laudominik committed:
drivers: dma: stm32: fix dma_stop stream busy handling for HAL override

For DMA channels overridden by HAL DMA, there was no way of
resetting stream busy variable.

Co-authored-by: Maciej Sobkowski <msobkowski@antmicro.com>
Signed-off-by: Dominik Lau <dlau@internships.antmicro.com>

laudominik committed:
drivers: disk: sdmmc: stm32: add missing DMA stream deinitialization

Due to deinitialization not happening, reinitialization
of the disk did not happen as the streams were marked busy.

Signed-off-by: Dominik Lau <dlau@internships.antmicro.com>

laudominik committed:
tests: drivers: dma: run dma tests on stm32f746g_disco

enabled dma loop transfer and channel blend to run
on stm32f746g_disco

Signed-off-by: Dominik Lau <dlau@internships.antmicro.com>

laudominik committed:
drivers: disk: sdmmc: enabled SDMMC DMA for stm32f7x

This enables the usage of DMA for stm32f7x in the sdmmc driver.

Co-authored-by: Maciej Sobkowski <msobkowski@antmicro.com>
Signed-off-by: Dominik Lau <dlau@internships.antmicro.com>

erwango reviewed (changes requested):
Thanks for the change, but I'm a bit puzzled that this is making 2 different flows for Zephyr dma API flow and HAL DMA API.
Can this be integrated in `stm32_sdmmc_dma_init` ? (another way could do as well as long as both HAL a,d API calls are kept in sync).

msobkowski commented:
> Thanks for the change, but I'm a bit puzzled that this is making 2 different flows for Zephyr dma API flow and HAL DMA API. Can this be integrated in `stm32_sdmmc_dma_init` ? (another way could do as well as long as both HAL a,d API calls are kept in sync).

Thanks for the comment @erwango , could you elaborate on how would you propose to handle it differently? Currently the DMA is initialized in the sm32_sdmmc driver using `dma_config()`, which will return an error if `stream->busy` is set, and for HAL override channels it is set to true in the `dma_stm32_configure()` function.

The filesystem sample recently got extended with remouting the disk which results in `dma_config()` being called twice, it works fine for the initial mount but for the second one it fails with:
```
[00:00:00.016,000] <err> dma_stm32: dma stream 6 is busy.
[00:00:00.016,000] <err> stm32_sdmmc: Failed to conig
[00:00:00.016,000] <err> stm32_sdmmc: failed to init tx dma
[00:00:00.016,000] <err> stm32_sdmmc: DMA init failed
[00:00:00.016,000] <err> fs: fs mount error (-5)
```
And this is what the change was attempting to address. If `dma_stop()` should not be called when using a HAL override channel than maybe the `stream->busy` check should be skipped for those channels in `dma_stm32_configure()`?

erwango commented:
@msobkowski I'm totally fine with the fact that `dma_stop()` should be called in order to enable SDMMC deinit/reinit.
The issue I'm pointing is that `HAL_DMA_DeInit();` is happening in `static int stm32_sdmmc_dma_init()` in a different step than the new function you're invoking.
What I'd like is to keep those close so that they're kept in sync.

msobkowski commented:
Hi @erwango, thanks for the clarification, I've pushed an updated version of the SDMMC deinitialization with `HAL_DMA_DeInit()` being called now in `stm32_sdmmc_dma_deinit()`, could you take another look?

msobkowski commented:
@erwango: Could you PTAL? Thanks!

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -608,8 +608,8 @@ DMA_STM32_EXPORT_API int dma_stm32_stop(const struct device *dev, uint32_t id)
 
-	/* Repeated stop : return now if channel is already stopped */
-	if (!stm32_dma_is_enabled_stream(dma, id)) {
+	/* Repeated stop : return now if channel is already stopped (if not overridden by the HAL)*/
+	if (!stream->hal_override && !stm32_dma_is_enabled_stream(dma, id)) {
 		return 0;
 	}
```
In case of `hal_override`, the only thing you need to do is to clear the busy flag.
So I'd suggest to do it directly before anything else and return directly.

erwango replied:
Note also that same should be done in dma_stm32u5 and dma_stm32_bdma drivers

msobkowski replied:
Fixed it in dma_stm32, applied the same change to the other two drivers.

erwango commented on drivers/disk/Kconfig.sdmmc:
```diff
@@ -2,7 +2,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
-DT_STM32_SDMMC_HAS_DMA := $(dt_nodelabel_has_prop,sdmmc,dmas)
+DT_STM32_SDMMC_HAS_DMA := $(dt_nodelabel_has_prop,sdmmc1,dmas)
```
Tbh, I'm not actually quite sure how this used to work, but this would obviously not work for sdmmc2 (for instance available on stm32f765).
`dt_compat_any_has_prop` would solve the problem

msobkowski replied:
Thanks for the tip, switched to use `dt_compat_any_has_prop` instead.

erwango commented on drivers/disk/Kconfig.sdmmc:
```diff
@@ -45,7 +45,7 @@ config SDMMC_STM32
 	select USE_STM32_LL_SDMMC
 	select USE_STM32_HAL_DMA if (SOC_SERIES_STM32L4X || SOC_SERIES_STM32F7X || SOC_SERIES_STM32F4X)
-	select DMA if ($(DT_STM32_SDMMC_HAS_DMA) && SOC_SERIES_STM32F4X)
+	select DMA if $(DT_STM32_SDMMC_HAS_DMA) && (SOC_SERIES_STM32L4X || SOC_SERIES_STM32F7X)
```
Not sure why the F4 disappeared here.

msobkowski replied:
That was a mistake, F4 got replaced with L4. Fixed it.

erwango reviewed (changes requested):
Some more tweaks required but direction is good.
Thanks for the changes so far.

msobkowski commented:
Tanks for the comments @erwango, I have addressed your remarks, could you take another look?

erwango reviewed (dismissed):
Thanks for this work!

msobkowski commented:
Thanks for the review. @danieldegrasse, @FRASTM: we need a second approval, could either of you take a look at this PR?

msobkowski committed:
samples: fs: Add stm32f746g_disco with DMA enabled to fs_sample

Adds custom overlay and config allowing the fs_sample to run on
the stm32f746g_disco with DMA enabled.

Signed-off-by: Maciej Sobkowski <msobkowski@antmicro.com>

msobkowski commented on tests/drivers/dma/chan_blen_transfer/boards/stm32f746g_disco.overlay:
```diff
@@ -0,0 +1,13 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&dma1 {
```
Only dma2 supports mem2mem transfers, as per the TRM:

>The DMA1 controller AHB peripheral port is not connected to the bus matrix like in the case of the DMA2 controller, thus only DMA2 streams are able to perform memory-to-memory transfers. 

Removed dma1 from the overlay.

msobkowski commented on tests/drivers/dma/chan_blen_transfer/boards/stm32f746g_disco.overlay:
Applied, thanks

msobkowski commented on tests/drivers/dma/loop_transfer/boards/stm32f746g_disco.overlay:
```diff
@@ -0,0 +1,13 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&dma1 {
```
Removed here as well, see comment above

msobkowski commented:
@danieldegrasse: could you take another look at this PR?

danieldegrasse reviewed (approved):
I'm fine with this approach- just something to be aware of, similar restrictions exist on other SD host controllers implemented using the SD subsystem. For those controllers, the subsystem will automatically copy unaligned data to an aligned buffer before passing it to the SD host controller. This change could also improve performance on those controllers by avoiding the copy though, so it's ok with me.
